src/types/acls.rs

```diff
@@ -0,0 +1,287 @@
+use crate::error::{Error, ParseError};
+use crate::types::UnsolicitedResponse;
+use crate::parse::try_handle_unilateral;
+use imap_proto::Response;
+use std::collections::HashSet;
+use std::sync::mpsc;
+
+// re-export some types from imap_proto
+pub use imap_proto::types::AclRight;
+
+/// enum used for set_acl to specify how the ACL is to be modified.
+pub enum AclModifyMode {
+    /// Replace all ACLs on the identifier for the mailbox
+    Replace,
+    /// Add the given ACLs to the identifier for the mailbox
+    Add,
+    /// Remove the given ACLs from the identifier for the mailbox
+    Remove,
+}
+
+/// Helpful wrapper around the ACL rights vector
+#[derive(Debug, Eq, PartialEq)]
+pub struct AclRightList {
+    pub(crate) data: HashSet<AclRight>,
+}
+
+impl AclRightList {
+    /// Returns if the AclRightList has the provided ACL (either as a char or an AclRight enum)
+    pub fn has_right<T: Into<AclRight>>(&self, right: T) -> bool {
+        self.data.contains(&right.into())
+    }
+}
+
+impl ToString for AclRightList {
+    fn to_string(&self) -> String {
+        let mut v: Vec<char> = self.data.iter().map(|c| char::from(*c)).collect();
+
+        v.sort_unstable();
+
+        v.into_iter().collect()
+    }
+}
+
+impl From<HashSet<AclRight>> for AclRightList {
+    fn from(hash: HashSet<AclRight>) -> Self {
+        Self { data: hash }
+    }
+}
+
+impl From<Vec<AclRight>> for AclRightList {
+    fn from(vec: Vec<AclRight>) -> Self {
+        AclRightList {
+            data: vec.into_iter().collect(),
+        }
+    }
+}
+
+impl From<&str> for AclRightList {
+    fn from(i: &str) -> Self {
+        i.chars()
+            .into_iter()
+            .map(|c| c.into())
+            .collect::<HashSet<AclRight>>()
+            .into()
+    }
+}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The ACL response from the getacl IMAP command
+#[derive(Debug, Eq, PartialEq)]
+pub struct Acl {
+    /// the mailbox these rights list are for
+    pub mailbox: String,
```

Comment:
This should almost certainly be `Cow<'a, str>` so that we can do zero-copy parsing and not require copying to use this in a request.

Reply:
I've added a new commit that does this..  It makes things more complex IMHO.
See the changes here: d901b36273cff675a566c6b36e10070e56280112

Reply:
It does, but it also adds a feature :)